src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -735,4 +730,324 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
+
+  /** Builder for HashSet.
+   */
+  private[collection] final class HashMapBuilder[A, B] extends mutable.ReusableBuilder[(A, B), HashMap[A, B]] {
+    import java.util
+
+    /* Nodes in the tree are either regular HashMap1, HashTrieMap, HashMapCollision1, or a mutable HashTrieMap
+      mutable HashTrieMap nodes are designated by having size == -1
+
+      mutable HashTrieMap nodes can have child nodes that are mutable, or immutable
+      immutable HashTrieMap child nodes can only be immutable
+
+      mutable HashTrieMap elems are always a Array of size 32,size -1, bitmap -1
+     */
+
+    /** The root node of the partially build hashmap */
+    private var rootNode: HashMap[A, B] = HashMap.empty
+    private def plusPlusMerger = liftMerger[A, B](null)
+    private def isMutable(hs: HashMap[A, B]) = {
+      hs.isInstanceOf[HashTrieMap[A, B]] && hs.size == -1
+    }
+
+    private def makeMutable(hs: HashTrieMap[A, B]): HashTrieMap[A, B] = {
+      if (isMutable(hs)) hs
+      else {
+        val elems = new Array[HashMap[A, B]](32)
+        var bit = 0
+        var iBit = 0
+        while (bit < 32) {
+          if ((hs.bitmap & (1 << bit)) != 0) {
+            elems(bit) = hs.elems(iBit)
+            iBit += 1
+          }
+          bit += 1
+        }
+        new HashTrieMap[A, B](-1, elems, -1)
+      }
+    }
+    private def isLeaf(hm: HashMap[A, B]) = {
+      hm.isInstanceOf[HashMap1[A, B]] || hm.isInstanceOf[HashMapCollision1[A, B]]
+    }
+    @inline private def computeHash(key: A) = computeHashImpl(key)
+
+    private def makeImmutable(hs: HashMap[A, B]): HashMap[A, B] = {
+      hs match {
+        case trie: HashTrieMap[A, B] if isMutable(trie) =>
+          var bit = 0
+          var bitmap = 0
+          var size = 0
+          while (bit < 32) {
+            if (trie.elems(bit) ne null)
+              trie.elems(bit) = makeImmutable(trie.elems(bit))
+            if (trie.elems(bit) ne null) {
+              bitmap |= 1 << bit
+              size += trie.elems(bit).size
+            }
+            bit += 1
+          }
+          Integer.bitCount(bitmap) match {
+            case 0 => null
+            case 1
+              if isLeaf(trie.elems(Integer.numberOfTrailingZeros(bitmap))) =>
+              trie.elems(Integer.numberOfTrailingZeros(bitmap))
+
+            case bc =>
+              val elems = if (bc == 32) trie.elems else {
+                val elems = new Array[HashMap[A, B]](bc)
+                var oBit = 0
+                bit = 0
+                while (bit < 32) {
+                  if (trie.elems(bit) ne null) {
+                    elems(oBit) = trie.elems(bit)
+                    oBit += 1
+                  }
+                  bit += 1
+                }
+                assert(oBit == bc)
+                elems
+              }
+              trie.size0 = size
+              trie.elems0 = elems
+              trie.bitmap0 = bitmap
+              trie
+          }
+        case _ => hs
+      }
```

Comment:
We should call `ScalaRuntime.releaseFence` (which will itself need to be backported but made `private[scala]`) here.

Reply:
@mkeskells  I just merged https://github.com/scala/scala/pull/8779 so you could now add `s.c.i.VM.releaseFence()` to this and #8722 in the same places as 2.13.x.